Mathlib/Data/Fin/Basic.lean

```diff
@@ -2570,14 +2571,16 @@ protected theorem zero_mul [NeZero n] (k : Fin n) : (0 : Fin n) * k = 0 := by
 end Mul
 
 open Qq in
-instance toExpr (n : ℕ) : Lean.ToExpr (Fin n) where
-  toTypeExpr := q(Fin $n)
-  toExpr := match n with
-    | 0 => finZeroElim
-    | k + 1 => fun i => show Q(Fin $n) from
-      have i : Q(Nat) := Lean.mkRawNatLit i -- raw literal to avoid ofNat-double-wrapping
-      have : Q(NeZero $n) := haveI : $n =Q $k + 1 := ⟨⟩; by exact q(NeZero.succ)
-      q(OfNat.ofNat $i)
+protected def toExprQ (i : Fin n) : Q(Fin $n) :=
```

Comment:
What's the motivation for giving this a standalone def? Can't we just use the typeclass-parametrized `toExprQ`?

Reply:
It would be possible to inline the definition here into the instance (and that's in fact how you should call it).  I usually try to make separate definitions for the operations to reduce the size of the instance after unfolding.  Whenever Lean reduces, for example, `ToExprQ.level`, it needs to copy the whole instance body, including both the `toTypeExprQ` definition and the even larger (when inlined) `toExprQ` definition.

(For other type classes like `Ring` there's an additional benefit:  When we check that two instances of, for example, `Add` are defeq, it is obviously much faster to check that `Fin.add =?= Fin.add` than to compare the definitions, which have modulo operators, additional type class arguments, etc.)

Writing proofs inside the instance is completely fine though, because Lean will automatically create auxiliary `_proof_7` lemmas for you.